okio/src/commonMain/kotlin/okio/FileHandle.kt

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+/**
+ * An open file for reading and writing; using either streaming and random access.
+ *
+ * Use [read] and [write] to perform one-off random-access reads and writes. Use [source], [sink],
+ * and [appendingSink] for streaming reads and writes.
+ *
+ * File handles must be closed when they are no longer needed. It is an error to read, write, or
+ * create streams after a file handle is closed. The operating system resources held by a file
+ * handle will be released once the file handle **and** all of its streams are closed.
+ *
+ * Although this class offers both reading and writing APIs, file handle instances may be
+ * read-only or write-only. For example, a handle to a file on a read-only file system will throw an
+ * exception if a write is attempted.
+ *
+ * File handles may be used by multiple threads concurrently. But the individual sources and sinks
+ * produced by a file handle are not safe for concurrent use.
+ */
+@ExperimentalFileSystem
+abstract class FileHandle : Closeable {
+  /**
+   * True once the file handle is closed. Resources should be released with [closeInternal] once
+   * this is true and [openStreamCount] is 0.
+   */
+  private var closed = false
+
+  /**
+   * Reference count of the number of open sources and sinks on this file handle. Resources should
+   * be released with [closeInternal] once this is 0 and [closed] is true.
+   */
+  private var openStreamCount = 0
+
+  /**
+   * Removes at least 1, and up to [byteCount] bytes from this and appends them to [sink]. Returns
+   * the number of bytes read, or -1 if this file is exhausted.
+   */
+  @Throws(IOException::class)
+  abstract fun read(offset: Long = 0L, sink: Buffer, byteCount: Long): Long
```

Comment:
Is there any alternative to Buffer here as the target that would allow for optimised operations internally within the filesystem implementation?

Reply:
If we wanna do such optimizations, we can change buffer to support ’em!

Reply:
I was wondering what a file to file direct copy would look like? Operations like https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#transferFrom(java.nio.channels.ReadableByteChannel,%20long,%20long)

Reply:
I wonder whether there is a good set of common operations to support? https://www.baeldung.com/java-filechannel

Reply:
That’s a good list. I think probably these:

- Reading
- Writing
- Current Position
- Get file size
- Set file size (ie. truncate)
- Force data
- Force metadata

And not these:
 - Memory mapping
 - Locks

These last two are exotic enough that omitting them shouldn’t be missed by most developers.

Reply:
As for direct filesystem transfer, that’s a neat optimization and it’s possible our APIs don’t preclude them!

But it’s a difficult API to use because it’s easy to fall off the fast path, and there’s no way to make sure that you’re on the fast path.

I’d prefer to omit that.

Reply:
Maybe add another case which is dropping into native APIs.  Make sure that you design for that case.  Here is how to bridge to java.nio, or node filesystem.

Reply:
To be clear, I'm not pushing for where the line is, just being clear around where it is and an image in your mind that it will generally work for typical client use cases, combined with the types of FileSystems you see implemented.